src/node/hooks/express/openapi.js

```diff
@@ -0,0 +1,666 @@
+              type: 'integer',
+            },
+          },
+        },
+      },
+      responses: {
+        ...defaultResponses,
+      },
+      securitySchemes: {
+        ApiKey: {
+          type: 'apiKey',
+          in: 'query',
+          name: 'apikey',
+        },
+      },
+    },
+    security: [{ ApiKey: [] }],
+  };
+
+  // build operations
+  for (const funcName in apiHandler.version[version]) {
+    let operation = {};
+    if (operations[funcName]) {
+      operation = { ...operations[funcName] };
+    } else {
+      // console.warn(`No operation found for function: ${funcName}`);
+      operation = {
+        operationId: funcName,
+        responses: defaultResponseRefs,
+      };
+    }
+
+    // set parameters
+    operation.parameters = operation.parameters || [];
+    for (const paramName of apiHandler.version[version][funcName]) {
+      operation.parameters.push({ $ref: `#/components/parameters/${paramName}` });
+      if (!definition.components.parameters[paramName]) {
+        definition.components.parameters[paramName] = {
+          name: paramName,
+          in: 'query',
+          schema: {
+            type: 'string',
+          },
+        };
+      }
+    }
+
+    // set path
+    let path = `/${operation.operationId}`; // APIPathStyle.FLAT
+    if (style === APIPathStyle.REST && operation._restPath) {
+      path = operation._restPath;
+    }
+    delete operation._restPath;
+
+    // add to definition
+    // NOTE: It may be confusing that every operation can be called with both GET and POST
+    definition.paths[path] = {
+      get: {
+        ...operation,
+        operationId: `${operation.operationId}UsingGET`,
+      },
+      post: {
+        ...operation,
+        operationId: `${operation.operationId}UsingPOST`,
+      },
+    };
+  }
+  return definition;
+};
+
+exports.expressCreateServer = async (_, args) => {
+  const { app } = args;
+
+  // create openapi-backend handlers for each api version under /api/{version}/*
+  for (const version in apiHandler.version) {
+    // we support two different styles of api: flat + rest
+    // TODO: do we really want to support both?
+
+    for (const style of [APIPathStyle.FLAT, APIPathStyle.REST]) {
+      const apiRoot = getApiRootForVersion(version, style);
+
+      // generate openapi definition for this API version
+      const definition = generateDefinitionForVersion(version, style);
+
+      // serve version specific openapi definition
+      app.get(`${apiRoot}/openapi.json`, (req, res) => {
+        res.header('Access-Control-Allow-Origin', '*');
+        res.json({ ...definition, servers: [generateServerForApiVersion(apiRoot, req)] });
+      });
+
+      // serve latest openapi definition file under /api/openapi.json
+      const isLatestAPIVersion = version === apiHandler.latestApiVersion;
+      if (isLatestAPIVersion) {
+        app.get(`/${style}/openapi.json`, (req, res) => {
+          res.header('Access-Control-Allow-Origin', '*');
```

Comment:
Same here.

`Access-Control-Allow-Origin: *` is convenient when developing, but wee cannot leave it hardcoded.

Reply:
The reasoning for doing this here was purely to maintain existing behaviour and a great deal of effort was put into this PR to not introduce any breaking changes.

In fact it looks like all API paths always return `Access-Control-Allow-Origin: *` at the moment: https://github.com/ether/etherpad-lite/blob/2b753f13ea3dac02132033b3ab0be0dce3931c19/src/node/hooks/express/swagger.js#L435-L437 

I agree that this is not good practice but I would still propose that we fix this as part of a separate PR.

Reply:
Woah. It's really there.

I agree: it is better for this series to be a refactoring in the strictest sense. Bug-for-bug, vuln-for-vuln.

Reply:
Looks like I was wrong. Only api paths starting with `/rest/{version}/**` seem to actually return `access-control-allow-origin: *` right now, even though `swagger-node-express` seems to clearly be configured to manage api paths. 

I'm very confused why this is but it looks like `/api/**` don't expose any cors headers in `develop`.

I'm unsure of what to do here. Should we just replicate this odd behaviour for this PR?

Reply:
I've just replicated the behaviour from `develop` branch for now. Feel free to push to this branch to change cors behaviour!

Commit: 4fda128